File: src/api/github.ts
import { invoke } from "@tauri-apps/api/core";

import type {
  GithubConfiguration,
  GithubEnvironment,
  GithubEnvironmentConfiguration,
  GithubIssueDetail,
  GithubIssueEdit,
  GithubOverview,
  GithubPullRequestDetail,
  GithubProject,
  GithubReleaseDetail,
  GithubReleaseEdit,
  GithubDiscussion,
  GithubCodespace,
  GithubDetailValue,
} from "./types";

export function getGithubOverview(path: string): Promise<GithubOverview> {
  return invoke<GithubOverview>("get_github_overview", { path });
}

export function getPullRequestDetail(
  path: string,
  number: number,
): Promise<GithubPullRequestDetail> {
  return invoke<GithubPullRequestDetail>("get_pull_request_detail", { path, number });
}

export function getIssueDetail(path: string, number: number): Promise<GithubIssueDetail> {
  return invoke<GithubIssueDetail>("get_issue_detail", { path, number });
}

export function getGithubConfiguration(path: string): Promise<GithubConfiguration> {
  return invoke<GithubConfiguration>("get_github_configuration", { path });
}

export function getGithubEnvironments(path: string): Promise<GithubEnvironment[]> {
  return invoke<GithubEnvironment[]>("get_github_environments", { path });
}

export function getGithubEnvironmentConfiguration(
  path: string,
  environment: string,
): Promise<GithubEnvironmentConfiguration> {
  return invoke<GithubEnvironmentConfiguration>("get_github_environment_configuration", {
    path,
    environment,
  });
}

export function saveGithubEnvironment(path: string, name: string): Promise<void> {
  return invoke("save_github_environment", { path, name });
}

export function deleteGithubEnvironment(path: string, name: string): Promise<void> {
  return invoke("delete_github_environment", { path, name });
}

export function setGithubEnvironmentVariable(
  path: string, environment: string, name: string, value: string,
): Promise<void> {
  return invoke("set_github_environment_variable", { path, environment, name, value });
}

export function deleteGithubEnvironmentVariable(
  path: string, environment: string, name: string,
): Promise<void> {
  return invoke("delete_github_environment_variable", { path, environment, name });
}

export function setGithubEnvironmentSecret(
  path: string, environment: string, name: string, value: string,
): Promise<void> {
  return invoke("set_github_environment_secret", { path, environment, name, value });
}

export function deleteGithubEnvironmentSecret(
  path: string, environment: string, name: string,
): Promise<void> {
  return invoke("delete_github_environment_secret", { path, environment, name });
}

export function setGithubVariable(path: string, name: string, value: string): Promise<void> {
  return invoke("set_github_variable", { path, name, value });
}

export function deleteGithubVariable(path: string, name: string): Promise<void> {
  return invoke("delete_github_variable", { path, name });
}

export function setGithubSecret(path: string, name: string, value: string): Promise<void> {
  return invoke("set_github_secret", { path, name, value });
}

export function deleteGithubSecret(path: string, name: string): Promise<void> {
  return invoke("delete_github_secret", { path, name });
}

export function saveGithubLabel(
  path: string,
  name: string,
  color: string,
  description: string,
): Promise<void> {
  return invoke("save_github_label", { path, name, color, description });
}

export function deleteGithubLabel(path: string, name: string): Promise<void> {
  return invoke("delete_github_label", { path, name });
}

export function checkGithubRulesets(path: string): Promise<string> {
  return invoke<string>("check_github_rulesets", { path });
}

export function getGithubRunLog(path: string, databaseId: number): Promise<string> {
  return invoke<string>("get_github_run_log", { path, databaseId });
}

export function rerunGithubRun(path: string, databaseId: number): Promise<void> {
  return invoke("rerun_github_run", { path, databaseId });
}

export function runGithubWorkflow(path: string, workflowId: number): Promise<void> {
  return invoke("run_github_workflow", { path, workflowId });
}

export function cancelGithubRun(path: string, databaseId: number): Promise<void> {
  return invoke("cancel_github_run", { path, databaseId });
}

export function downloadGithubArtifacts(
  path: string,
  databaseId: number,
  targetDir: string,
  artifactName = "",
): Promise<void> {
  return invoke("download_github_artifacts", {
    path,
    databaseId,
    targetDir,
    artifactName,
  });
}

export function createGithubRelease(
  path: string,
  tag: string,
  title: string,
  notes: string,
  draft = false,
  prerelease = false,
): Promise<string> {
  return invoke<string>("create_github_release", {
    path,
    tag,
    title,
    notes,
    draft,
    prerelease,
  });
}

export function getReleaseDetail(path: string, tag: string): Promise<GithubReleaseDetail> {
  return invoke<GithubReleaseDetail>("get_release_detail", { path, tag });
}

export function editGithubRelease(
  path: string,
  tag: string,
  edit: GithubReleaseEdit,
): Promise<void> {
  return invoke("edit_github_release", { path, tag, edit });
}

export function uploadGithubReleaseAssets(
  path: string,
  tag: string,
  files: readonly string[],
  clobber = false,
): Promise<void> {
  return invoke("upload_github_release_assets", { path, tag, files, clobber });
}

export function downloadGithubRelease(
  path: string,
  tag: string,
  targetDir: string,
  pattern = "",
): Promise<void> {
  return invoke("download_github_release", { path, tag, targetDir, pattern });
}

export function forkGithubRepository(path: string, organization = ""): Promise<string> {
  return invoke<string>("fork_github_repository", { path, organization });
}

export function syncGithubRepository(path: string, branch = ""): Promise<void> {
  return invoke("sync_github_repository", { path, branch });
}

export function cloneGithubRepository(reference: string, targetDir: string): Promise<string> {
  return invoke<string>("clone_github_repository", { reference, targetDir });
}

export function createGithubRepository(
  path: string,
  name: string,
  visibility: "public" | "private" | "internal",
  description = "",
): Promise<string> {
  return invoke<string>("create_github_repository", {
    path,
    name,
    visibility,
    description,
  });
}

export function createPullRequest(
  path: string,
  title: string,
  body: string,
  draft = false,
): Promise<string> {
  return invoke<string>("create_pull_request", { path, title, body, draft });
}

export function reviewPullRequest(
  path: string,
  number: number,
  action: "approve" | "comment" | "request-changes",
  body: string,
): Promise<void> {
  return invoke("review_pull_request", { path, number, action, body });
}

export function mergePullRequest(
  path: string,
  number: number,
  method: "merge" | "squash" | "rebase" = "squash",
  deleteBranch = true,
): Promise<void> {
  return invoke("merge_pull_request", { path, number, method, deleteBranch });
}

export function createIssue(path: string, title: string, body: string): Promise<string> {
  return invoke<string>("create_issue", { path, title, body });
}

export function editIssue(path: string, number: number, edit: GithubIssueEdit): Promise<void> {
  return invoke("edit_issue", { path, number, edit });
}

export function commentIssue(path: string, number: number, body: string): Promise<void> {
  return invoke("comment_issue", { path, number, body });
}

export function closeIssue(path: string, number: number): Promise<void> {
  return invoke("close_issue", { path, number });
}

export function getGithubProjects(path: string): Promise<GithubProject[]> {
  return invoke<GithubProject[]>("get_github_projects", { path });
}

export function getGithubProjectItems(
  path: string, number: number, query = "",
): Promise<GithubDetailValue> {
  return invoke<GithubDetailValue>("get_github_project_items", { path, number, query });
}

export function getGithubDiscussions(path: string): Promise<GithubDiscussion[]> {
  return invoke<GithubDiscussion[]>("get_github_discussions", { path });
}

export function getGithubCodespaces(path: string): Promise<GithubCodespace[]> {
  return invoke<GithubCodespace[]>("get_github_codespaces", { path });
}

export function getGithubCodespaceLog(path: string, name: string): Promise<string> {
  return invoke<string>("get_github_codespace_log", { path, name });
}

export function stopGithubCodespace(path: string, name: string): Promise<void> {
  return invoke("stop_github_codespace", { path, name });
}

export function deleteGithubCodespace(
  path: string, name: string, force: boolean,
): Promise<void> {
  return invoke("delete_github_codespace", { path, name, force });
}

export function searchGithub(
  path: string,
  kind: "code" | "commits" | "issues" | "prs" | "repos",
  query: string,
  currentRepository = false,
): Promise<GithubDetailValue[]> {
  return invoke<GithubDetailValue[]>("search_github", {
    path,
    kind,
    query,
    currentRepository,
  });
}

export function runSafeGithubCommand(
  path: string, command: string, extraArgs: readonly string[],
): Promise<string> {
  return invoke<string>("run_safe_github_command", { path, command, extraArgs });
}

export function runGithubApiRequest(
  path: string,
  method: "GET" | "POST" | "PUT" | "PATCH" | "DELETE",
  endpoint: string,
  fields: readonly { key: string; value: string }[],
): Promise<string> {
  return invoke<string>("run_github_api_request", { path, method, endpoint, fields });
}

File: src-tauri/src/commands/github.rs
use std::path::PathBuf;

use crate::github::{
    self, GithubConfiguration, GithubIssueDetail, GithubIssueEdit, GithubOverview,
    GithubPullRequestDetail,
};

#[tauri::command]
pub async fn get_github_overview(path: PathBuf) -> Result<GithubOverview, String> {
    tauri::async_runtime::spawn_blocking(move || github::overview(&path))
        .await
        .map_err(|error| error.to_string())?
}

#[tauri::command]
pub async fn get_pull_request_detail(
    path: PathBuf,
    number: u64,
) -> Result<GithubPullRequestDetail, String> {
    tauri::async_runtime::spawn_blocking(move || github::pull_request_detail(&path, number))
        .await
        .map_err(|error| error.to_string())?
}

#[tauri::command]
pub async fn get_issue_detail(path: PathBuf, number: u64) -> Result<GithubIssueDetail, String> {
    tauri::async_runtime::spawn_blocking(move || github::issue_detail(&path, number))
        .await
        .map_err(|error| error.to_string())?
}

#[tauri::command]
pub async fn get_github_configuration(path: PathBuf) -> Result<GithubConfiguration, String> {
    tauri::async_runtime::spawn_blocking(move || github::configuration(&path))
        .await
        .map_err(|error| error.to_string())?
}

#[tauri::command]
pub async fn set_github_variable(path: PathBuf, name: String, value: String) -> Result<(), String> {
    tauri::async_runtime::spawn_blocking(move || github::set_variable(&path, &name, &value))
        .await
        .map_err(|error| error.to_string())?
}

#[tauri::command]
pub async fn delete_github_variable(path: PathBuf, name: String) -> Result<(), String> {
    tauri::async_runtime::spawn_blocking(move || github::delete_variable(&path, &name))
        .await
        .map_err(|error| error.to_string())?
}

#[tauri::command]
pub async fn set_github_secret(path: PathBuf, name: String, value: String) -> Result<(), String> {
    tauri::async_runtime::spawn_blocking(move || github::set_secret(&path, &name, &value))
        .await
        .map_err(|error| error.to_string())?
}

#[tauri::command]
pub async fn delete_github_secret(path: PathBuf, name: String) -> Result<(), String> {
    tauri::async_runtime::spawn_blocking(move || github::delete_secret(&path, &name))
        .await
        .map_err(|error| error.to_string())?
}

#[tauri::command]
pub async fn save_github_label(
    path: PathBuf,
    name: String,
    color: String,
    description: String,
) -> Result<(), String> {
    tauri::async_runtime::spawn_blocking(move || {
        github::save_label(&path, &name, &color, &description)
    })
    .await
    .map_err(|error| error.to_string())?
}

#[tauri::command]
pub async fn delete_github_label(path: PathBuf, name: String) -> Result<(), String> {
    tauri::async_runtime::spawn_blocking(move || github::delete_label(&path, &name))
        .await
        .map_err(|error| error.to_string())?
}

#[tauri::command]
pub async fn check_github_rulesets(path: PathBuf) -> Result<String, String> {
    tauri::async_runtime::spawn_blocking(move || github::check_default_branch_rules(&path))
        .await
        .map_err(|error| error.to_string())?
}

#[tauri::command]
pub async fn get_github_run_log(path: PathBuf, database_id: u64) -> Result<String, String> {
    tauri::async_runtime::spawn_blocking(move || github::run_log(&path, database_id))
        .await
        .map_err(|error| error.to_string())?
}

#[tauri::command]
pub async fn rerun_github_run(path: PathBuf, database_id: u64) -> Result<(), String> {
    tauri::async_runtime::spawn_blocking(move || github::rerun(&path, database_id))
        .await
        .map_err(|error| error.to_string())?
}

#[tauri::command]
pub async fn run_github_workflow(path: PathBuf, workflow_id: u64) -> Result<(), String> {
    tauri::async_runtime::spawn_blocking(move || github::run_workflow(&path, workflow_id))
        .await
        .map_err(|error| error.to_string())?
}

#[tauri::command]
pub async fn cancel_github_run(path: PathBuf, database_id: u64) -> Result<(), String> {
    tauri::async_runtime::spawn_blocking(move || github::cancel_run(&path, database_id))
        .await
        .map_err(|error| error.to_string())?
}

#[tauri::command]
pub async fn download_github_artifacts(
    path: PathBuf,
    database_id: u64,
    target_dir: String,
    artifact_name: String,
) -> Result<(), String> {
    tauri::async_runtime::spawn_blocking(move || {
        github::download_run_artifacts(&path, database_id, &target_dir, &artifact_name)
    })
    .await
    .map_err(|error| error.to_string())?
}

#[tauri::command]
pub async fn create_github_release(
    path: PathBuf,
    tag: String,
    title: String,
    notes: String,
    draft: bool,
    prerelease: bool,
) -> Result<String, String> {
    tauri::async_runtime::spawn_blocking(move || {
        github::create_release(&path, &tag, &title, &notes, draft, prerelease)
    })
    .await
    .map_err(|error| error.to_string())?
}

#[tauri::command]
pub async fn download_github_release(
    path: PathBuf,
    tag: String,
    target_dir: String,
    pattern: String,
) -> Result<(), String> {
    tauri::async_runtime::spawn_blocking(move || {
        github::download_release(&path, &tag, &target_dir, &pattern)
    })
    .await
    .map_err(|error| error.to_string())?
}

#[tauri::command]
pub async fn fork_github_repository(path: PathBuf, organization: String) -> Result<String, String> {
    tauri::async_runtime::spawn_blocking(move || github::fork_repository(&path, &organization))
        .await
        .map_err(|error| error.to_string())?
}

#[tauri::command]
pub async fn sync_github_repository(path: PathBuf, branch: String) -> Result<(), String> {
    tauri::async_runtime::spawn_blocking(move || github::sync_repository(&path, &branch))
        .await
        .map_err(|error| error.to_string())?
}

#[tauri::command]
pub async fn clone_github_repository(
    reference: String,
    target_dir: String,
) -> Result<String, String> {
    tauri::async_runtime::spawn_blocking(move || github::clone_repository(&reference, &target_dir))
        .await
        .map_err(|error| error.to_string())?
}

#[tauri::command]
pub async fn create_github_repository(
    path: PathBuf,
    name: String,
    visibility: String,
    description: String,
) -> Result<String, String> {
    tauri::async_runtime::spawn_blocking(move || {
        github::create_repository(&path, &name, &visibility, &description)
    })
    .await
    .map_err(|error| error.to_string())?
}

#[tauri::command]
pub async fn create_pull_request(
    path: PathBuf,
    title: String,
    body: String,
    draft: bool,
) -> Result<String, String> {
    tauri::async_runtime::spawn_blocking(move || {
        github::create_pull_request(&path, &title, &body, draft)
    })
    .await
    .map_err(|error| error.to_string())?
}

#[tauri::command]
pub async fn review_pull_request(
    path: PathBuf,
    number: u64,
    action: String,
    body: String,
) -> Result<(), String> {
    tauri::async_runtime::spawn_blocking(move || {
        github::review_pull_request(&path, number, &action, &body)
    })
    .await
    .map_err(|error| error.to_string())?
}

#[tauri::command]
pub async fn merge_pull_request(
    path: PathBuf,
    number: u64,
    method: String,
    delete_branch: bool,
) -> Result<(), String> {
    tauri::async_runtime::spawn_blocking(move || {
        github::merge_pull_request(&path, number, &method, delete_branch)
    })
    .await
    .map_err(|error| error.to_string())?
}

#[tauri::command]
pub async fn create_issue(path: PathBuf, title: String, body: String) -> Result<String, String> {
    tauri::async_runtime::spawn_blocking(move || github::create_issue(&path, &title, &body))
        .await
        .map_err(|error| error.to_string())?
}

#[tauri::command]
pub async fn edit_issue(path: PathBuf, number: u64, edit: GithubIssueEdit) -> Result<(), String> {
    tauri::async_runtime::spawn_blocking(move || github::edit_issue(&path, number, &edit))
        .await
        .map_err(|error| error.to_string())?
}

#[tauri::command]
pub async fn comment_issue(path: PathBuf, number: u64, body: String) -> Result<(), String> {
    tauri::async_runtime::spawn_blocking(move || github::comment_issue(&path, number, &body))
        .await
        .map_err(|error| error.to_string())?
}

#[tauri::command]
pub async fn close_issue(path: PathBuf, number: u64) -> Result<(), String> {
    tauri::async_runtime::spawn_blocking(move || github::close_issue(&path, number))
        .await
        .map_err(|error| error.to_string())?
}
